### PR TITLE
fix: add multi-layer fuzzy dedup to prevent duplicate observations

### LINKED ISSUE #1260: Duplicate observations still occurring after PR #1214 fix - concurrent hook triggers bypass content-hash deduplication
## Problem

PR #1214 was supposed to fix duplicate observations (#598), but duplicates are still being created in v10.5.2.

## Root Cause

The content-hash deduplication in #1214 only prevents **identical content** from being stored twice. However, the actual duplication happens because **multiple hooks fire concurrently for a single tool call**, each independently calling the AI to summarize the same event. Since AI generates slightly different wording each time, every observation gets a unique `content_hash`, bypassing deduplication entirely.

## Evidence

Database query on v10.5.2 shows:

| ID | Title | Timestamp | content_hash |
|----|-------|-----------|--------------|
| 625 | 检查 claude-mem GitHub 最新发布信息 | 2026-02-27T03:56:54.470Z | DA1984F0... |
| 626 | 检查 claude-mem GitHub 最新发布信息 | 2026-02-27T03:56:54.470Z | EEEF2FB9... |
| 627 | 检查 claude-mem GitHub 最新发布信息 | 2026-02-27T03:56:54.470Z | 855808D3... |
| ... | (same title, same timestamp) | ... | (all different hashes) |
| 635 | 检查 claude-mem GitHub 最新发布信息 | 2026-02-27T03:56:54.470Z | BB211827... |

10-11 identical observations created at the **exact same millisecond** with different hashes.

## Analysis

- Same timestamp → triggered by concurrent hook invocations, not separate events
- Different `content_hash` → AI generates unique wording each time → dedup check passes → all get stored
- PR #1214 fix works for **replay/retry** duplicates but not **concurrent hook fan-out**

## Expected Fix

Deduplication should be based on **tool call identity** (same tool + same input + same session within N seconds), not content hash. This would prevent the N concurrent hooks from all writing separate observations for the same underlying event.

## Environment

- claude-mem version: 10.5.2
- OS: macOS 14
- Database: SQLite at ~/.claude-mem/claude-mem.db

### PR DESCRIPTION
**Pull request by tyu-mit**
## Summary

LLMs (especially Gemini) generate near-identical observation blocks that bypass exact content-hash matching, leading to duplicate observations in the database. This PR adds three defense-in-depth layers to prevent duplicates:

- **Parser-level**: deduplicate within a single LLM response using Levenshtein similarity (threshold 0.95) before observations reach storage
- **Hash normalization**: apply Unicode NFC + whitespace collapsing before hashing so invisible differences produce the same content hash
- **DB-level fuzzy fallback**: when hash match fails, check for same-title observations within the dedup window and compare narrative similarity

Resolves #1260

## Test plan

- [x] Unit tests for string similarity utility (`tests/utils/string-similarity.test.ts`)
- [x] Parser dedup tests (`tests/sdk/parser-dedup.test.ts`)
- [x] Data integrity tests for normalization and fuzzy dedup (`tests/sqlite/data-integrity.test.ts`)
- [ ] Manual verification: run with Gemini provider and confirm duplicate observations are eliminated

### CODE REVIEW

File: plugin/scripts/context-generator.cjs [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
...AND created_at_epoch > ? LIMIT 5").all(s,o);for(let m of d){let p=ge(m.narrative??"",n??"");if(p===0&&(m.narrative??"").length>1e3){u.debug("DEDUP",`Fuzzy match skipped \u2014 narrative exceeds max length for Levenshtein | title="${s}" | existingId=${m.id}`);continue}if(p>xt)return u.debug("DEDUP",`Fuzzy match found | title="${s}" | similarity=${p.toFixed(3)} | existingId=${m.id}`),{id:m.id,created_at_epoch:m.created_at_epoch}}}return null}var F=class{db;constructor(e=ue){e!==":memory:"&&_e(C),this.db=new be.Database(e),this.db.run("PRAGMA journal_mode = WAL"),this.db.run("PRAGMA synchronous = NORMAL"),this.db.run("PRAGMA foreign_keys = ON"),this.initializeSchema(),this.ensureWorkerPortColumn(),this.ensurePromptTrackingColumns(),this.removeSessionSummariesUniqueConstraint(),this.addObservationHierarchicalFields(),this.makeObservationsTextNullable(),this.createUserPromptsTable(),this.ensureDiscoveryTokensColumn(),this.createPendingMessagesTable(),this.renameSessionIdColumns(),this.repairSessionIdColumnRename(),this.addFailedAtEpochColumn(),this.addOnUpdateCascadeToForeignKeys(),this.addObservationContentHashColumn(),this.addSessionCustomTitleColumn()}initializeSchema(){this.db.run(`
```
[the remark this answers is not in the document]
Comment (tyu-mit) [reply plugin/scripts/context-generator.cjs]: **Design decision — not changing.**

The exact title match (`WHERE title = ?`) is a deliberate conservative choice. The hash layer already catches exact dupes; the fuzzy layer catches same-title-different-narrative (LLM wording variations). Broadening to fetch recent observations regardless of title would change performance characteristics and semantics significantly — better suited as a separate enhancement PR if needed.

File: plugin/scripts/context-generator.cjs [hunk 2]
```diff
@@ -435,84 +435,84 @@ ${o.stack}`:` ${o.message}`:this.getLevel()===0&&typeof o=="object"?_=`
...
+    `).get(e,t)?.prompt_text??null}storeObservation(e,t,s,n,o=0,i){let a=i??Date.now(),d=new Date(a).toISOString(),m=w(e,s.title,s.narrative),p=$(this.db,m,a);if(p)return{id:p.id,createdAtEpoch:p.created_at_epoch};let g=this.db.prepare(`
       INSERT INTO observations
       (memory_session_id, project, type, title, subtitle, facts, narrative, concepts,
        files_read, files_modified, prompt_number, discovery_tokens, content_hash, created_at, created_at_epoch)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
-    `).run(e,t,s.type,s.title,s.subtitle,JSON.stringify(s.facts),s.narrative,JSON.stringify(s.concepts),JSON.stringify(s.files_read),JSON.stringify(s.files_modified),n||null,o,p,d,a);return{id:Number(T.lastInsertRowid),createdAtEpoch:a}}storeSummary(e,t,s,n,o=0,i){let a=i??Date.now(),d=new Date(a).toISOString(),_=this.db.prepare(`
+    `).run(e,t,s.type,s.title,s.subtitle,JSON.stringify(s.facts),s.narrative,JSON.stringify(s.concepts),JSON.stringify(s.files_read),JSON.stringify(s.files_modified),n||null,o,m,d,a);return{id:Number(g.lastInsertRowid),createdAtEpoch:a}}storeSummary(e,t,s,n,o=0,i){let a=i??Date.now(),d=new Date(a).toISOString(),p=this.db.prepare(`
```
[the remark this answers is not in the document]
Comment (tyu-mit) [reply plugin/scripts/context-generator.cjs]: **Fixed.** Singular `storeObservation` in both the TypeScript source (`SessionStore.ts:1528`) and the bundled CJS now passes `title` and `narrative` to `findDuplicateObservation()`. See commit 0eaea03.

File: src/sdk/parser.ts [hunk 3]
```diff
@@ -95,6 +99,35 @@ export function parseObservations(text: string, correlationId?: string): ParsedO
...
+  // narrativeSim > 0.95), unlike storage-level dedup in findDuplicateObservation()
+  // which requires exact title match + fuzzy narrative. Parser handles within-response
+  // duplicates; storage handles cross-request duplicates (concurrent agents, retries).
+  if (observations.length > 1) {
+    const deduped: ParsedObservation[] = [];
+    for (const obs of observations) {
+      const isDuplicate = deduped.some(existing => {
+        const titleSim = similarity(existing.title ?? '', obs.title ?? '');
+        const narrativeSim = similarity(existing.narrative ?? '', obs.narrative ?? '');
+        return titleSim > PARSER_DEDUP_THRESHOLD && narrativeSim > PARSER_DEDUP_THRESHOLD;
```
[the remark this answers is not in the document]
Comment (tyu-mit) [reply src/sdk/parser.ts]: **Fixed.** Added early return when either `title` or `narrative` is missing/empty on either side of the comparison. Also added a test case (`preserves observations with empty/null title or narrative`). See commit 0eaea03.

File: src/services/sqlite/observations/store.ts [hunk 4]
```diff
@@ -6,41 +6,95 @@
...
+ * (trailing spaces, multiple spaces, Unicode variants) from producing different hashes.
  */
 export function computeObservationContentHash(
   memorySessionId: string,
   title: string | null,
   narrative: string | null
 ): string {
   return createHash('sha256')
-    .update((memorySessionId || '') + (title || '') + (narrative || ''))
+    .update((memorySessionId || '') + normalizeForHash(title) + normalizeForHash(narrative))
```
[the remark this answers is not in the document]
Comment (tyu-mit) [reply src/services/sqlite/observations/store.ts]: **Fixed.** Changed to null-byte (`\0`) delimited join: `[memorySessionId, normalizeForHash(title), normalizeForHash(narrative)].join('\0')`. See commit 0eaea03.

File: src/services/sqlite/observations/store.ts [hunk 5]
```diff
@@ -6,41 +6,95 @@
...
+ *
  * Returns the existing observation's id and timestamp if found, null otherwise.
  */
 export function findDuplicateObservation(
   db: Database,
   contentHash: string,
-  timestampEpoch: number
+  timestampEpoch: number,
+  title?: string | null,
+  narrative?: string | null
```
[the remark this answers is not in the document]
Comment (tyu-mit) [reply src/services/sqlite/observations/store.ts]: **Fixed.** `findDuplicateObservation()` now accepts `memorySessionId` as its second parameter. Both the hash query and the fuzzy title query include `WHERE memory_session_id = ?`. All 6 call sites updated (3 in `SessionStore.ts`, 2 in `transactions.ts`, 1 in `store.ts`). See commit 0eaea03.

File: src/services/sqlite/SessionStore.ts [hunk 6]
```diff
@@ -1670,9 +1670,9 @@ export class SessionStore {
       `);
 
       for (const observation of observations) {
-        // Content-hash deduplication (same logic as storeObservation singular)
+        // Content-hash deduplication with fuzzy fallback (same logic as storeObservation singular)
         const contentHash = computeObservationContentHash(memorySessionId, observation.title, observation.narrative);
-        const existing = findDuplicateObservation(this.db, contentHash, timestampEpoch);
+        const existing = findDuplicateObservation(this.db, contentHash, timestampEpoch, observation.title, observation.narrative);
```
[the remark this answers is not in the document]
Comment (tyu-mit) [reply src/services/sqlite/SessionStore.ts]: **Fixed.** Singular `storeObservation` now passes all arguments: `findDuplicateObservation(this.db, memorySessionId, contentHash, timestampEpoch, observation.title, observation.narrative)`. See commit 0eaea03.

**Comment by tyu-mit** [comment]
Thanks for the thorough review @xkonjin! Addressing your questions:

> **1. Levenshtein perf with 1KB limit**

The `MAX_LEVENSHTEIN_LENGTH = 1000` guard in `string-similarity.ts:15` returns `0.0` immediately for any string > 1KB, skipping the O(n*m) computation entirely. Combined with `LIMIT 5` on candidates, worst case is 5 comparisons of ≤1KB strings — negligible latency.

> **2. New params in fuzzy fallback**

Yes — `title` and `narrative` are used in the secondary (fuzzy) path of `findDuplicateObservation()`. The batch methods already passed them correctly; the singular `storeObservation` path was missing them (good catch from CodeRabbit). Fixed in commit 0eaea03.

> **3. Configurable threshold**

Implemented in 0eaea03! Added `CLAUDE_MEM_DEDUP_SIMILARITY_THRESHOLD` to settings (default `0.95`). Configurable via `~/.claude-mem/settings.json` or environment variable. Both parser-level and storage-level dedup read from it. Added tests covering threshold override behavior.

File: src/sdk/parser.ts [hunk 7]
```diff
@@ -95,6 +103,39 @@ export function parseObservations(text: string, correlationId?: string): ParsedO
...
+    const deduped: ParsedObservation[] = [];
+    const threshold = getDedupThreshold();
+    for (const obs of observations) {
+      const isDuplicate = deduped.some(existing => {
+        if (!existing.title || !obs.title || !existing.narrative || !obs.narrative) {
+          return false;
+        }
+        const titleSim = similarity(existing.title, obs.title);
+        const narrativeSim = similarity(existing.narrative, obs.narrative);
+        return titleSim > threshold && narrativeSim > threshold;
```
[the remark this answers is not in the document]
Comment (tyu-mit) [reply src/sdk/parser.ts]: Fixed in 7dee1bc — changed `>` to `>=` in both parser and storage layers, and clamped `getDedupThreshold()` return value to `[0, 1]` with `Number.isFinite` + `Math.min/max`.

File: src/services/worker/http/routes/SettingsRoutes.ts [hunk 8]
```diff
@@ -122,6 +122,8 @@ export class SettingsRoutes extends BaseRouteHandler {
       'CLAUDE_MEM_CONTEXT_SHOW_LAST_SUMMARY',
       'CLAUDE_MEM_CONTEXT_SHOW_LAST_MESSAGE',
       'CLAUDE_MEM_FOLDER_CLAUDEMD_ENABLED',
+      // Deduplication
+      'CLAUDE_MEM_DEDUP_SIMILARITY_THRESHOLD',
```
[the remark this answers is not in the document]
Comment (tyu-mit) [reply src/services/worker/http/routes/SettingsRoutes.ts:128]: Fixed in 7dee1bc — added validation in `validateSettings()` that rejects non-finite values and values outside `[0, 1]` range with a clear error message.

**Comment by tyu-mit** [comment]
Thanks @xkonjin\! The threshold is already configurable via `CLAUDE_MEM_DEDUP_SIMILARITY_THRESHOLD` in `~/.claude-mem/settings.json` (default `0.95`). Added input validation for it in 7dee1bc as well — rejects values outside `[0, 1]` range.

**Comment by thedotmack** [comment]
Closed during the April 2026 backlog cleanup. The underlying bug is now tracked in #1955, which is the single canonical issue being addressed by the maintainer. Thanks for taking the time to report — your symptoms and repro are captured in the consolidated ticket.
